### PR TITLE
rpc: Handle `getinfo` client-side in bitcoin-cli w/ `-getinfo`

### PR DESCRIPTION
**Pull request by laanwj**
(follow-up to #8780)

There have been some requests for a client-side `getinfo` and this is my PoC of how to do it. The high-level idea is to have one command to gather and show various information and statistics about a `bitcoind` in one go.

This could be used as an opportunity to make `getinfo` better, it doesn't have to have exactly the same fields as the original RPC, e.g. a `chain` instead of `testnet` flag would make sense (but it is currently the same, according to this table: https://github.com/bitcoin/bitcoin/issues/8455#issuecomment-250468383). 

If this is considered a good idea some of the logic could be moved up to rpcclient.cpp and used in the GUI console as well.
### Implementation

This adds the infrastructure `BaseRequestHandler` class that takes care of converting bitcoin-cli arguments into a JSON-RPC request object, and converting the reply into a JSON object that can be shown as result.

This is subsequently used to handle the `-getinfo` option, which sends a JSON-RPC batch request to the RPC server with `["getnetworkinfo", "getblockchaininfo", "getwalletinfo"]`, and after reply combines the result into what looks like a `getinfo` result.

### CODE REVIEW

**Comment by sipa** [comment]
Concept ACK. This allows us to maintain backward compatibility with "bitcoin-cli getinfo" without keeping it at the RPC level. Should "-getinfo" automatically trigger when the getinfo RPC is requested?

**Comment by laanwj** [comment]
> Should "-getinfo" automatically trigger when the getinfo RPC is requested?

Not sure about that. That's how I had it first but this would blur the line between what bitcoin-cli calls an RPC call and what is actually available on the API endpoint.

Maybe "getinfo" if it fails (with the code that the call is missing) could print an helpful message to use "-getinfo" instead.

**Comment by luke-jr** [comment]
Eh, IMO the only backward-compatibility need for getinfo is at the RPC level. RPC isn't really designed for end users, and in any case, end users can adapt to changes easier than legacy software making RPC calls.

I don't particularly care about whether -getinfo is added to bitcoin-cli, but it doesn't seem like a viable migration path or substitute for the getinfo RPC.

**Comment by laanwj** [comment]
> I don't particularly care about whether -getinfo is added to bitcoin-cli, but it doesn't seem like a viable migration path or substitute for the getinfo RPC.

You are right that RPC is not there for end users.
However, bitcoin-cli is there for end users, and many people use bitcoind+bitcoin-cli as a console wallet.

This is not just for backwards-compatibility. As I say in my OP

> The high-level idea is to have one command to gather and show various information and statistics about a bitcoind in one go.

I'd be fine with completely different output here. However, the idea of having a single call that combines and prints various status information makes sense, IMO.

**Comment by laanwj** [comment]
And yes, this could be an external Python script too. I don't really care. Although history has shown that those scripts hardly get maintained nor used, e.g. remember `contrib/bitrpc`? Thought so.

**Comment by laanwj** [comment]
Reopening after IRC discussion.

**Comment by laanwj** [comment]
rebased

File: src/rpc/protocol.cpp [hunk 1]
```diff
@@ -124,3 +124,19 @@ void DeleteAuthCookie()
     }
 }
 
+std::vector<UniValue> JSONRPCProcessBatchReply(const UniValue &in, size_t num)
+{
+    if (!in.isArray())
```
Comment (jnewbery) [inline src/rpc/protocol.cpp]: nit: new code so should use braces for `if` blocks (sorry!)

**Review by jnewbery — COMMENTED** [review]
Looks good.

>> Should "-getinfo" automatically trigger when the getinfo RPC is requested?

> That's how I had it first but this would blur the line between what bitcoin-cli calls an RPC call and what is actually available on the API endpoint.

I think `bitcoin-cli getinfo` *should* continue to work. I would also lean towards not using the `bitcoin-cli -getinfo` form. Yes, it's a bit magic, but it would mean maximum compatibility with any scripts, etc that are calling `bitcoin-cli getinfo`, and there'd be no excuse to not remove the actual RPC. If you're feeling really fastidious, you could add an extra `"warnings"` field to the returned object to say that getinfo isn't a real RPC. The returned object already has an `"errors"` field, so that wouldn't look too strange.

It's a shame that we don't have support in `test/functional` for bitcoin-cli tests. I don't think it's worth adding that just for this new function, but longer term we should aim to add tests.

Just one piece of functional feedback, and a couple of nits.

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -189,7 +190,93 @@ static void http_error_cb(enum evhttp_request_error err, void *ctx)
...
+        }
+        result.pushKV("blocks", batch[ID_BLOCKCHAININFO]["result"]["blocks"]);
+        result.pushKV("timeoffset", batch[ID_NETWORKINFO]["result"]["timeoffset"]);
+        result.pushKV("connections", batch[ID_NETWORKINFO]["result"]["connections"]);
+        result.pushKV("proxy", batch[ID_NETWORKINFO]["result"]["networks"][0]["proxy"]);
+        result.pushKV("difficulty", batch[ID_BLOCKCHAININFO]["result"]["difficulty"]);
+        result.pushKV("testnet", UniValue(batch[ID_BLOCKCHAININFO]["result"]["chain"].get_str() == "test"));
+        if (!batch[ID_WALLETINFO].isNull()) {
+            result.pushKV("keypoololdest", batch[ID_WALLETINFO]["result"]["keypoololdest"]);
+            result.pushKV("keypoolsize", batch[ID_WALLETINFO]["result"]["keypoolsize"]);
```
Comment (jnewbery) [inline src/bitcoin-cli.cpp]: Actually, I've got this backwards. `getinfo` should be updated to be consistent with `getwalletinfo`, not making `-getinfo` consistent with the incorrect `getinfo`.

I think that can be done as a separate commit in this PR. Let me know if you want me to make a branch with that commit.

**Comment by jnewbery** [comment]
I'd love for this to make it in to v0.15.0, to ease the removal of `getinfo` in (hopefully) v0.15.1.

**Comment by TheBlueMatt** [comment]
I am concerned that this breaks the atomicity of the previously-atomic RPC getinfo, which may be very surprising for users. I'm not sure how much it matters if you get inconsistent data in results from bitcoin-cli, since most automated things will likely be using RPC directly, but I'm not convinced it is a great direction.

**Comment by promag** [comment]
NACK. Keep it simple for cli. IMO this is a good example for what to not do in a cli and can lead to others do the same in their cli.. This doesn't break atomicity as there is no getinfo, at the very least that should be a concern in #8843. The ones requesting getinfo should either just make the calls or understand what they really need.

**Comment by jnewbery** [comment]
> I'm not convinced it is a great direction

> this is a good example for what to not do in a cli and can lead to others do the same in their cli

Idealogical purity is fine in its place, but this is an isolated change to help those who are currently use `getinfo` from the command line. I don't think it sets a precedent for doing more hacks with the cli.

Perhaps don't document this? Or document it as a debug option? Would that be better?

To be clear: I'm not advocating strongly for this PR. If the consensus view is that this is not required to remove `getinfo` then I'm happy to go along with that. However, I'm certainly not concerned about it setting precedent and I don't think that a good reason to not merge this.

**Comment by TheBlueMatt** [comment]
I'm having trouble parsing your comment @promag. My concern with atomicity is that you could eg get a getwalletinfo result which is current, then a getblockchaininfo result a block later and end up with a very confused getinfo. Forcing the client to make separate calls clarifies that somewhat.

Anyway, I'm curious what the previous IRC discussion was, let's bring this up at next week's meeting (or on IRC on Monday) and figure out what it is that the use-case is.

**Comment by laanwj** [comment]
Closing this, I don't feel like addressing the atomicity concern, it seems arbitrary, never came up before, and a way to call multiple RPC calls atomically is *WAY* out of scope for this. Not going to maintain this anymore.

**Comment by promag** [comment]
> but this is an isolated change to help those who are currently use getinfo from the command line

@jnewbery that's what deprecation is for.

> Forcing the client to make separate calls clarifies that somewhat.

@TheBlueMatt I totally agree with you! Sorry if I wasn't clear.

@laanwj IMO best solution is to add support for http://www.jsonrpc.org/specification#batch, where the locks are held while handling the batch request.

**Comment by laanwj** [comment]
I use a batch in the implementation of this!

Holding all the locks for the whole time seems a bad idea though. A batch
might include, or fully consist of, RPCs that don't need any locks at all.
Locks have been pushed down, just a version ago, to be able to handle them
separately per call implementation.

It could be done optionally; e.g. submit a list of locks with the batch.
But wtf, I had never before this even heard of a use case for that, seems
very rare.

**Comment by laanwj** [comment]
Anyhow I'm no longer going to work on this after the so-manyeth time this
gets shot down for out-of-the-blue reasons.

I don't even use it myself. The getinfo output doesn't provide a useful
selection of information. I have a custom script that prints what I'm
interested in, and suggest everyone does the same, easy enough.

**Comment by promag** [comment]
> I use a batch in the implementation of this!

@laanwj sorry I've overlook the code and did not realize there is already support for server side batch requests.

:+1: for custom scripting.

**Pull request closed without merging**